drivers/generic/generic.go

```diff
@@ -0,0 +1,221 @@
+package generic
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/codegangsta/cli"
+	"github.com/docker/machine/drivers"
+	"github.com/docker/machine/log"
+	"github.com/docker/machine/provider"
+	"github.com/docker/machine/state"
+	"github.com/docker/machine/utils"
+)
+
+type Driver struct {
+	MachineName    string
+	IPAddress      string
+	SSHKey         string
+	SSHUser        string
+	SSHPort        int
+	CaCertPath     string
+	PrivateKeyPath string
+	DriverKeyPath  string
+	SwarmMaster    bool
+	SwarmHost      string
+	SwarmDiscovery string
+	storePath      string
+}
+
+const (
+	defaultTimeout = 5 * time.Second
+)
+
+func init() {
+	drivers.Register("generic", &drivers.RegisteredDriver{
+		New:            NewDriver,
+		GetCreateFlags: GetCreateFlags,
+	})
+}
+
+// GetCreateFlags registers the flags this driver adds to
+// "docker hosts create"
+func GetCreateFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "generic-ip-address",
+			Usage: "IP Address of machine",
+			Value: "",
+		},
+		cli.StringFlag{
+			Name:  "generic-ssh-user",
+			Usage: "SSH user",
+			Value: "root",
+		},
+		cli.StringFlag{
+			Name:  "generic-ssh-key",
+			Usage: "SSH private key path",
+			Value: filepath.Join(utils.GetHomeDir(), ".ssh", "id_rsa"),
+		},
+		cli.IntFlag{
+			Name:  "generic-ssh-port",
+			Usage: "SSH port",
+			Value: 22,
+		},
+	}
+}
+
+func NewDriver(machineName string, storePath string, caCert string, privateKey string) (drivers.Driver, error) {
+	return &Driver{MachineName: machineName, storePath: storePath, CaCertPath: caCert, PrivateKeyPath: privateKey}, nil
+}
+
+func (d *Driver) DriverName() string {
+	return "generic"
+}
+
+func (d *Driver) AuthorizePort(ports []*drivers.Port) error {
+	return nil
+}
+
+func (d *Driver) DeauthorizePort(ports []*drivers.Port) error {
+	return nil
+}
+
+func (d *Driver) GetMachineName() string {
+	return d.MachineName
+}
+
+func (d *Driver) GetSSHHostname() (string, error) {
+	return d.GetIP()
+}
+
+func (d *Driver) GetSSHKeyPath() string {
+	return filepath.Join(d.storePath, "id_rsa")
+}
+
+func (d *Driver) GetSSHPort() (int, error) {
+	if d.SSHPort == 0 {
+		d.SSHPort = 22
+	}
+
+	return d.SSHPort, nil
+}
+
+func (d *Driver) GetSSHUsername() string {
+	if d.SSHUser == "" {
```

Comment:
In what situation would these be empty?  I can't think of why the user would set them that way.